### PR TITLE
fix aliasing for primtorch view meta kernels

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/86284

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86611
* #86610
* #86609
* #86228
* #86608
* #86607
* __->__ #86285

### CODE REVIEW

**Comment by mruberry** [comment]
The approach looks reasonable to me. What about extending `test_python_ref_meta` to detect this:

https://github.com/pytorch/pytorch/blob/17addb307ee9a4d12ad6918e90358a9a47a4f12b/test/test_ops.py#L174

**Comment by IvanYashchuk** [comment]
Is the only difference here that with `alias=True` the resulting tensor would return True for `tensor._is_view()`?
```py
import torch
a = torch.zeros(3, 2, device="meta")
b = torch._prims.TensorMeta(a, shape=a.shape, strides=a.stride())
print(b._is_view()) # False
c = a.as_strided(a.shape, a.stride())
print(c._is_view()) # True
assert a.storage().data_ptr() == b.storage().data_ptr() == c.storage().data_ptr() # It's all 0 for meta
```

**Comment by bdhirsh** [comment]
@mruberry Thanks for the pointer to `test_python_ref_meta`. It looks like today, that test currently checks the output shapes match, but [not strides or storage offsets](https://github.com/pytorch/pytorch/blob/cebf08afb24dec0720935b9a9bd64ecf05b472d5/torch/_prims_common/__init__.py#L126). The two ways I could imagine updating that test are:

(1) Update the test to check for the `_is_view()` relationship (this seems a bit weird since it's a private API that's dependent on autograd, but maybe this would be fine?) 
(1) updating the test to always check that storage offsets match. This is the "broken" thing that I originally noticed. That would mean that the tests always check that sizes and storage_offset match, but no strides. Does that sound reasonable to you?


FWIW, I'm also adding a cross ref test that tests correctness of all of the python decomps and meta functions in `torch/_decomp/decompositions.py` and `torch/_meta_registrations.py`. Some of those decomps call into prims, which is where the problem first showed up. Although those two files probably don't comprehensively call into the prims, so a separate prim test sounds good to me.

**Comment by mruberry** [comment]
> @mruberry Thanks for the pointer to `test_python_ref_meta`. It looks like today, that test currently checks the output shapes match, but [not strides or storage offsets](https://github.com/pytorch/pytorch/blob/cebf08afb24dec0720935b9a9bd64ecf05b472d5/torch/_prims_common/__init__.py#L126). The two ways I could imagine updating that test are:
> 
> (1) Update the test to check for the `_is_view()` relationship (this seems a bit weird since it's a private API that's dependent on autograd, but maybe this would be fine?) 

Maybe `TestViewOps` in `test_view_ops.py` can provide some inspiration. See

https://github.com/pytorch/pytorch/blob/cebf08afb24dec0720935b9a9bd64ecf05b472d5/test/test_view_ops.py#L96


> (1) updating the test to always check that storage offsets match. This is the "broken" thing that I originally noticed. That would mean that the tests always check that sizes and storage_offset match, but no strides. Does that sound reasonable to you?

Stride-testing has been disabled for the moment. I spent a good amount of time trying to emulate PyTorch's striding logic, but PyTorch is pretty inconsistent in how it handles strides itself. See https://github.com/pytorch/pytorch/issues/78050. @ezyang concludes that we'd like to emulate strides, but I'm not sure how we would do this.


> FWIW, I'm also adding a cross ref test that tests correctness of all of the python decomps and meta functions in `torch/_decomp/decompositions.py` and `torch/_meta_registrations.py`. Some of those decomps call into prims, which is where the problem first showed up. Although those two files probably don't comprehensively call into the prims, so a separate prim test sounds good to me.

There are several existing tests in `test_ops.py` for Python reference consistency, see

https://github.com/pytorch/pytorch/blob/cebf08afb24dec0720935b9a9bd64ecf05b472d5/test/test_ops.py#L339

These might be interesting to look at when developing a consistency test for decompositions. Ideally I think we'd like to see all decompositions ported to become Python references.

**Comment by ezyang** [comment]
@bdhirsh, why don't you just convert these into direct `as_strided` calls on the input tensor, rather than going through the TensorMeta constructor?

**Comment by bdhirsh** [comment]
> @bdhirsh, why don't you just convert these into direct `as_strided` calls on the input tensor, rather than going through the TensorMeta constructor?

Sounds good - you're right, seems cleaner

File: torch/_prims/__init__.py [hunk 1]
```diff
@@ -1173,7 +1173,7 @@ def _greater_than_reduce(acc, x):
         else:
             new_strides.append(0)
 
-    return TensorMeta(a, shape=shape, strides=new_strides)
+    return a.as_strided(shape, new_strides)
```
Comment (albanD) [inline torch/_prims/__init__.py:1176]: All the functions here need a `storage_offset=a.storage_offset()` argument!

**Review by albanD — COMMENTED** [review]
The storage_offset argument is missing. Sounds good otherwise.

**Review by albanD — APPROVED** [review]
SGTM

**Review by mruberry — APPROVED** [review]
LGTM!

**Pull request closed without merging**